🌱 Allow incremental condition patch

fabriziopandini opened the pull request:
**What this PR does / why we need it**:
This PR allows incremental condition patching within the same controller.

In order to do so, in the final defer patch, it is required to declare the list of conditions the controller owns, and for which, a conflict between after and latest value of the condition should not generate errors (after value should always be used instead)

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/3260

/assign @vincepri 
/assign @gab-satchi

vincepri commented on util/patch/options.go:
```diff
@@ -27,13 +29,28 @@ type HelperOptions struct {
+
+// IgnoreConditionsConflicts allows to define condition types for which the value defined in the object
+// passed to patch should take precedence on the latest value in case of conflicts.
+type IgnoreConditionsConflicts struct {
```
```suggestion
type WithOwnedConditions struct {
```

fabriziopandini replied:
Fixed!

fabriziopandini commented:
As a reminder pull-cluster-api-apidiff output

```
sigs.k8s.io/cluster-api/util/patch
  Incompatible changes:
  - HelperOptions: old is comparable, new is not
  Compatible changes:
  - HelperOptions.OwnedConditions: added
  - WithOwnedConditions: added

sigs.k8s.io/cluster-api/util/conditions
  Incompatible changes:
  - Patch.Apply: changed from func(Setter) error to func(Setter, ...ApplyOption) error
  - PatchOperation.Base: removed
  - PatchOperation.Target: removed
  Compatible changes:
  - ApplyOption: added
  - PatchOperation.After: added
  - PatchOperation.Before: added
  - WithOwnedConditions: added
```

Above changes are ok because those methods where not part of a release (they were included in v0.3.7 alpha 0, but this is fine)

vincepri commented:
/milestone v0.3.7

vincepri commented:
/assign @ncdc 

for final lgtm

gab-satchi commented on util/conditions/patch.go:
```diff
@@ -72,67 +72,114 @@ func NewPatch(before Getter, after Getter) Patch {
-				if !hasSameState(sourceCondition, conditionPatch.Base) {
-					return errors.Errorf("error patching conditions: The condition %q on was modified by a different process and this caused a merge/RemoveCondition conflict", conditionPatch.Base.Type)
+			// if so then this is a conflict (unless we should ignore conflicts for this condition type).
+			if sourceCondition := Get(latest, conditionPatch.Before.Type); sourceCondition != nil {
```
rename to `latestCondition` to match other cases

gab-satchi commented on util/conditions/patch.go:
```diff
@@ -72,67 +72,114 @@ func NewPatch(before Getter, after Getter) Patch {
+			// if so then this is a conflict (unless we should ignore conflicts for this condition type).
+			if sourceCondition := Get(latest, conditionPatch.Before.Type); sourceCondition != nil {
+				if !hasSameState(sourceCondition, conditionPatch.Before) {
+					return errors.Errorf("error patching conditions: The condition %q on was modified by a different process and this caused a merge/RemoveCondition conflict", conditionPatch.Before.Type)
```
```suggestion
					return errors.Errorf("error patching conditions: The condition %q was modified by a different process and this caused a merge/RemoveCondition conflict", conditionPatch.Before.Type)
```

gab-satchi commented on util/conditions/patch.go:
```diff
@@ -72,67 +72,114 @@ func NewPatch(before Getter, after Getter) Patch {
-			if sourceCondition == nil {
-				return errors.Errorf("error patching conditions: The condition %q on was deleted by a different process and this caused a merge/ChangeCondition conflict", conditionPatch.Target.Type)
+			if latestCondition == nil {
+				return errors.Errorf("error patching conditions: The condition %q on was deleted by a different process and this caused a merge/ChangeCondition conflict", conditionPatch.After.Type)
```
```suggestion
				return errors.Errorf("error patching conditions: The condition %q was deleted by a different process and this caused a merge/ChangeCondition conflict", conditionPatch.After.Type)
```

gab-satchi commented on util/conditions/patch.go:
```diff
@@ -72,67 +72,114 @@ func NewPatch(before Getter, after Getter) Patch {
+			// the after state corresponds to the desired value. If not this is a conflict (unless we should ignore conflicts for this condition type).
+			if !reflect.DeepEqual(latestCondition, conditionPatch.Before) {
+				if !hasSameState(latestCondition, conditionPatch.After) {
+					return errors.Errorf("error patching conditions: The condition %q on was modified by a different process and this caused a merge/ChangeCondition conflict", conditionPatch.After.Type)
```
```suggestion
					return errors.Errorf("error patching conditions: The condition %q was modified by a different process and this caused a merge/ChangeCondition conflict", conditionPatch.After.Type)
```

gab-satchi commented:
/lgtm
